src/main/java/at/hannibal2/skyhanni/data/PetAPI.kt

```diff
@@ -35,31 +193,425 @@ object PetAPI {
+                break
+            }
+        }
+        if (newPetLine == pet?.rawPetName || newPetLine == null) return
+
+        var petItem = NEUInternalName.NONE
+        var petXP: Double? = null
+        event.lines.forEach { line ->
+            val tempPetItem = handleWidgetStringLine(line)
+            if (tempPetItem != NEUInternalName.NONE) {
+                petItem = tempPetItem
+                return@forEach
+            }
+            val tempPetXP = handleWidgetXPLine(line)
+            if (tempPetXP != null) {
+                petXP = tempPetXP
+                return@forEach
+            }
+        }
+
+        event.lines.forEach { line ->
+            if (handleWidgetPetLine(line, petItem, petXP, newPetLine)) return@forEach
+        }
+    }
+
+    private fun handleWidgetPetLine(line: String, petItem: NEUInternalName, petXP: Double?, newPetLine: String): Boolean {
+        val xpOverLevel = petXP ?: 0.0
+        petWidgetPattern.matchMatcher(line) {
+            val xp = (levelToXP(
+                group("level").toInt(),
+                LorenzRarity.getByColorCode(group("rarity")[0]) ?: LorenzRarity.ULTIMATE,
+                group("name")
+            ))
+            val petName = group("name")
+            val rarity = LorenzRarity.getByColorCode(group("rarity")[0]) ?: LorenzRarity.ULTIMATE
+
+            val newPet = PetData(
+                petNameToInternalName(petName, rarity),
+                petName,
+                rarity,
+                petItem,
+                group("level").toInt(),
+                (xp?.plus(xpOverLevel)) ?: 0.0,
+                newPetLine,
+            )
+            fireEvent(newPet)
+            return true
+        }
+        return false
+    }
+
+    private fun handleWidgetStringLine(line: String): NEUInternalName {
+        widgetStringPattern.matchMatcher(line) {
+            val string = group("string")
+            if (string == "No pet selected") {
+                PetChangeEvent(pet, null).post()
+                pet = null
+                return NEUInternalName.NONE
+            }
+            return NEUInternalName.fromItemNameOrNull(string) ?: NEUInternalName.NONE
+        }
+        return NEUInternalName.NONE
+    }
+
+    private fun handleWidgetXPLine(line: String): Double? {
+        xpWidgetPattern.matchMatcher(line) {
+            if (group("max") != null) return null
+
+            val overflow = group("overflow")?.formatDoubleOrNull() ?: 0.0
+            val currentXP = group("currentXP")?.formatDoubleOrNull() ?: 0.0
+
+            return overflow + currentXP
+        }
+        return null
+    }
+
+    @SubscribeEvent
+    fun onChat(event: LorenzChatEvent) {
+        if (autopetMessagePattern.matches(event.message)) {
+            val hoverMessage = buildList {
+                event.chatComponent.hover?.siblings?.forEach {
+                    add(it.formattedText)
+                }
+            }.toString().split("\n")
```

Comment:
i really dont like calling toString on the list here
```suggestion
            val hoverMessage = event.chatComponent.hover?.siblings?.joinToString("")?.split("\n") ?: return
```